README.md
redis-memoizer
===

An asynchronous function memoizer for node.js, using redis as the memo store. Memos expire after a specified timeout. Great as a drop-in performance optimization / caching layer for heavy asynchronous functions.

Wikipedia [explains it best](http://en.wikipedia.org/wiki/Memoization):
> ...memoization is an optimization technique used primarily to speed up computer programs by having function calls avoid repeating the calculation of results for previously processed inputs.

```javascript
const redisClient = require('redis').createClient(port, host, options)
const memoize = require("redis-memoizer")(redisClient);

async function someExpensiveOperation(arg1, arg2) { /* ... */}

const memoized = memoize(someExpensiveOperation);
```

Now, calls to `memoized` will have the same effect as calling `someExpensiveOperation`, except it will be much faster. The results of the first call are stored in redis and then looked up for subsequent calls.

Redis effectively serves as a shared network-available cache for function calls. Thus, the memoization cache is available across processes, so that if the same function call is made from different processes they will reuse the cache.

## Changes

Version 2 is Promise-first and no longer supports callbacks. If you want to support a callback function,
promisify it first.

Version 2 is similar to the `master` branch of https://github.com/errorception/redis-memoizer but introduces
a few extra options, gzip support, support for reifying Errors and Dates and assumes promises everywhere. The
locking mechanism is also less strict and gives up sooner, assuming it is not catastrophic to calculate an expensive
fn more often if it reduces overall latency.

## Uses

Lets say you are making a DB call that's rather expensive. Let's say you've wrapped the call into a `getUserProfile` function that looks as follows:

```javascript
function getUserProfile(userId) {
	// Go over to the DB, perform expensive call, get user's profile
	return User.get(userId); // some promise
}
```

Let's say this call takes 500ms, which is unacceptably high, and you want to make it faster, and don't care about the fact that the value of `userProfile` might be slightly outdated (until the cache timeout is hit in redis). You could simply do the following:

```javascript
const getMemoizedUserProfile = memoize(getUserProfile);

// First call. This will take some time.
const userProfile = await getMemoizedUserProfile("user1");


// Second call. This will be blazingly fast.
const userProfile2 = await getMemoizedUserProfile("user1");

```

This can similarly be used for any network or disk bound async calls where you are tolerant of slightly outdated values.

## Usage

### Initialization

Creates a memoization function. Requires an existing redis client.

```javascript
const memoize = require("redis-memoizer")(redisClient, {
	// Properties prefixed with `default_` can be overridden when creating each memoized function.
	// How long to persist memoized results to Redis. This can be overridden per-fn.
	default_ttl: 120000,
	// How long to wait for the lock (including retries)
	default_lock_timeout: 5000,

	// How long to wait on Redis before just moving on.
  // If the TTL fed to `memoize` is shorter than this, it will be used instead.
	lookup_timeout: 1000, // ms
	// Set to a function that determines whether or not to memoize an error. By default, we never do.
	memoize_errors_when: (err) => false,
	// Namespace to use under `memos:` in redis. Useful to e.g. invalidate all cache after
  // an application update; simply set this to the current git revision or use the boot timestamp.
	memoize_key_namespace: null,
	// How often to spin on the lock
	lock_retry_delay: 50,

	// This is *required*, below is an example
	on_error: (err, client, key) => console.error(err),

	// Optional customizations. You don't have to touch these, but in certain
	// cases it might be useful to. They're provided here so you don't have
	// to fork the module to make these changes.

	// Serialize a value to a string before sending to redis.
	serialize_value(value, options) { ... },
	// Given a string, deserialize back to a JS object.
	deserialize_value(value, options) { ... },
	// For convenience, the default (de)serialization methods reference these
	// when dealing with errors. They are easy to override if you have
	// custom attributes, like e.g. `statusCode`.
	// 'message' is always present.
	error_serialization_keys: ['name', 'stack']

});
```

### memoize(asyncFunction: Function, options: {ttl?: number | (result: any?) => number, lock_timeout?: number, name: string})

Memoizes an async function and returns it.

* `asyncFunction` must be an asynchronous function that needs to be memoized. The function should return a Promise.

* `name` (`string`) *Required* identifier for this function. As of v2 this is required as we have found no perfect way to uniquely
identify a function (many functions look alike, like Promisify wrappers), while making the identifier consistent across processes.
Make sure this is unique across your application.

* `ttl` (`?number | (result?) => number`) (Default: 120000 (ms)) is the amount of time in milliseconds for which the result of the function call should be cached in redis. Once the timeout is hit, the value is deleted from redis automatically. This is done using the redis [`psetex` command](http://redis.io/commands/psetex). The timeout is only set the first time, so the value expires after the timeout time has expired since the first call. The timeout is not reset with every call to the memoized function. Once the value has expired in redis, this module will treat the function call as though it's called the first time again. `ttl` can alternatively be a function, if you want to dynamically determine the cache time based on the data returned. The returned data will be passed into the timeout function, but not always (for calculating minimums).

* `lock_timeout` (`?number`) This is how long to check spinlock when there isn't a result in Redis. See "Cache Stampedes" below. By default, this is 50ms. You may want to set
this lower for faster response times when multiple identical functions are called at the same time, but it will increase the load on Redis.


```javascript
const httpCallMemoized = memoize(makeHttpCall);
const res = await httpCallMemoized(options);
```

## Cache Stampedes

This module makes some effort to minimize the effect of a [cache stampede](http://en.wikipedia.org/wiki/Cache_stampede). If multiple calls are made in quick succession before the first (async) call has completed, only the first call is actually really made. Note that redis will not have been populated at this time yet. Subsequent calls are queued up and are responded to as soon as the result of the first call is available.

Once all the calls have been responded to and the result of the computation is stored in redis, the module then switches to using the computed values from redis.

Note, cache stampedes can still happen if the same function is called from different processes, since the queueing logic described above happens in-memory. For the same set of arguments, you are likely to make as many calls as you have processes.

## Types

Note that this module does serialization to JSON. Special affordances are made for Date objects, which will be correctly returned
as Dates, but other, more complex types (like Functions) will not survive the serialization/deserialization.

## Installation

Use npm to install redis-memoizer:
```
npm install redis-memoizer
```

To run the tests, install the dev-dependencies by `cd`'ing into `node_modules/redis-memoizer` and running `npm install` once, and then `npm test`.

## License

(The MIT License)

Copyright (c) 2012 Rakesh Pai <rakeshpai@errorception.com>

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# 5.0.0 (Apr 5, 2018)

* Breaking: Errors no longer serialize `type`
* Breaking: `on_error` is no longer called for TimeoutErrors
* Feature: Added `serialize_value(value: any, options: Options) => string` and `deserialize_value(value: string, options: Options) => any` options if you need deep customization.
* Feature: Added `error_serialization_keys` convenience array. Use if you have custom error keys such as `statusCode`.
* Refactored some functions and exports.

# 4.0.0 (Jan 30, 2018)

* Breaking: Remove `error_logging` option and move to required `on_error` function with arity `(err: Error, client: RedisClient, key: string) => void`.

# 3.0.0 (Jan 19, 2018)

* Breaking: `error_logging` option is now a function of arity `(client: RedisClient, key: string) => boolean` rather than a bool.

# 2.1.2 (Dec 24, 2017)

* Fix: Lookup timeouts could return `undefined`

# 2.1.1 (Dec 24, 2017)

* Fix: Fix handling of Buffers in ioredis

# 2.1.0 (Dec 24, 2017)

* Feature: Add node_redis and fakeredis compatibility in library and test.

# 2.0.0 (Dec 24, 2017)

* Breaking: Library now uses Promises everywhere, not callbacks.
* Breaking: `name` option is now required when memoizing functions. Automatically determining function uniqueness is impractical across multiple running processes. Provide a unique name instead and the memoizer will use it as a key. This helps us reliably share results across many services.
* Breaking: Move compatibility to Node >= 8.
* Feature: New locking implementation.

# 0.6.0 (Oct 10, 2017)

* Feature: ioredis support

# 0.5.2 (July 28, 2017)

* Feature: Export uuid & hash functions
* Performance: Avoid assigning properties to input functions to avoid conversion to slow object
* Fix: Memory leak on `inFlight`
* Test: Use `fakeredis`

# 0.5.1 (June 22, 2017)

* Fix: Error when gzipping exotic objects

# 0.5.0 (Mar 08, 2017)

* Breaking: drop Node < 4 support
* Refactor: Use a proper uuid library

> See the commit history for prior revisions.


test/redis-memoizer.js
const crypto = require('crypto');
const should = require('should');
const Promise = require('bluebird');

const memoizePkg = require('../');
const pkgJSON = JSON.stringify(require('../package.json'));
const {exec} = require('../redisCompat');
const PORT = parseInt(process.env.PORT) || 6379;

const hash = (string) => crypto.createHmac('sha1', 'memo').update(string).digest('hex');
const key_namespace = Date.now();
const TYPS = {
  ioredis: 'ioredis',
  node_redis: 'redis',
  fakeredis: 'fakeredis',
};

async function delKeys(client, keyPattern) {
  const keys = await exec(client, 'keys', keyPattern);
  return await Promise.map(keys, (key) => exec(client, 'del', key.toString()));
}

function makeDefaultOptions() {
  return {
    memoize_key_namespace: key_namespace,
    on_error: (err) => { throw err; },
  };
}


describe('redis-memoizer', () => {

  const REDIS_TYP = TYPS[process.env.REDIS_TYP] || 'ioredis';
  const Redis = require(REDIS_TYP);
  if (REDIS_TYP !== 'ioredis') {
    Promise.promisifyAll(Redis.RedisClient.prototype);
    Promise.promisifyAll(Redis.Multi.prototype);
  }
  const client = Redis.createClient(PORT, 'localhost', {return_buffers: true});
  if (REDIS_TYP === 'fakeredis') client.options.return_buffers = true; // workaround check, it sets opts badly

  let memoize = memoizePkg(client, {
    ...makeDefaultOptions(),
  });

  const originalMemoize = memoize;
  // Record the name so we can clear this out
  memoize = function(fn, options) {
    // Ensures we can't clobber our own names
    fn._name = String(Math.random()) + memoizePkg.getFunctionKey(fn, options);
    return originalMemoize.apply(this, arguments);
  };

  // Wait for connection
  before(async () => {
    if (REDIS_TYP === 'ioredis' && client.status !== 'connect') {
      return new Promise((resolve) => client.on('connect', resolve));
    }
  });

	afterEach(async () => {
    await delKeys(client, `memos:${key_namespace}:*`);
  });

  after(() => {
		client.end(true);
  });

  it('should memoize a value correctly', async () => {
    const functionDelayTime = 10;
    let callCount = 0;
    const functionToMemoize = async (val1, val2) => {
      callCount++;
      await Promise.delay(functionDelayTime);
      return { val1, val2 };
    };
    const memoized = memoize(functionToMemoize, {name: 'fn1'});

    let start = Date.now();
    let { val1, val2 } = await memoized(1, 2);
    val1.should.equal(1);
    val2.should.equal(2);
    (Date.now() - start >= functionDelayTime).should.be.true;		// First call should go to the function itself
    callCount.should.equal(1);

    start = Date.now();
    ({ val1, val2 } = await memoized(1, 2));
    val1.should.equal(1);
    val2.should.equal(2);
    (Date.now() - start < functionDelayTime).should.be.true;		// Second call should be faster
    callCount.should.equal(1);
  });

  it('should memoize large values (gzip)', async () => {
    let callCount = 0;
    const functionToMemoize = async (val1) => {
      callCount++;
      return pkgJSON + val1;
    };
    const memoized = memoize(functionToMemoize, {name: 'pkg'});

    const json = await memoized(1);
    json.should.equal(pkgJSON + 1);
    callCount.should.equal(1);

    const json2 = await memoized(1);
    json2.should.equal(pkgJSON + 1);
    callCount.should.equal(1);
  });

  it('should memoize separate function separately', async () => {
    const function1 = async arg => { await Promise.delay(10); return 1; };
    const function2 = async arg => { await Promise.delay(10); return 2; };

    const memoizedFn1 = memoize(function1, {name: 'fn1'});
    const memoizedFn2 = memoize(function2, {name: 'fn2'});

    (await memoizedFn1('x')).should.equal(1);
    (await memoizedFn2('y')).should.equal(2);
    (await memoizedFn1('x')).should.equal(1);
  });

  it('should memoize same function with different args separately', async () => {
    const fn = async arg => { await Promise.delay(10); return arg; };

    const memoizedFn = memoize(fn, {name: 'fn_args'});

    (await memoizedFn('x')).should.equal('x');
    (await memoizedFn('y')).should.equal('y');
    (await memoizedFn('x')).should.equal('x');
  });

  async function stampede(options, memoizerOptions, memoizeOptions) {
    // this thing is crazy slow, breaks tests
    if (REDIS_TYP === 'fakeredis') return memoizerOptions.lock_retry_delay * 10;

    if (!options.delayTime || !options.iters) throw new Error('bad args', options);
    let callCount = 0;
    const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
      ...memoizerOptions,
    });

    const fn = async () => {
      callCount++;
      await Promise.delay(options.delayTime);
    };
    const memoized = do_memoize(fn, memoizeOptions);

    let start = Date.now();
    await Promise.all([ ...Array(options.iters).keys() ].map(() => memoized()));
    let duration = Date.now() - start;
    callCount.should.equal(1);
    return duration;
  }

  it('should prevent a cache stampede (slow lock)', async () => {
    const options = {delayTime: 10, iters: 10};
    const duration = await stampede(options, {lock_retry_delay: 50}, {name: 'fn_lock_slow'});
    (duration).should.be.above(options.delayTime * options.iters);
  });

  it('should prevent a cache stampede (fast lock)', async () => {
    const options = {delayTime: 10, iters: 10};
    const duration = await stampede(options, {lock_retry_delay: 1}, {name: 'fn_lock_fast'});
    (duration).should.be.below(options.delayTime * options.iters);
  });

  it('should prevent a cache stampede (low lock timeout)', async () => {
    const options = {delayTime: 10, iters: 10};
    const duration = await stampede(options, {lock_retry_delay: 1}, {name: 'fn_lock_fast2', lock_timeout: 20});
    (duration).should.be.below(50);
  });

  it(`should respect 'this'`, async () => {
    function Obj() { this.x = 1; }
    Obj.prototype.y = async function() {
      await Promise.delay(10);
      return this.x;
    };

    const obj = new Obj();
    const memoizedY = memoize(obj.y, {name: 'fn_this'}).bind(obj);

    (await memoizedY()).should.equal(1);
  });

  it('should respect the ttl', async () => {
    const ttl = 100;
    const functionDelayTime = 10;

    const fn = () => Promise.delay(functionDelayTime);
    const memoized = memoize(fn, {name: 'fn1', ttl});

    let start = Date.now();
    await memoized();
    (Date.now() - start >= functionDelayTime).should.be.true;

    // Call immediately again. Should be a cache hit.
    start = Date.now();
    await memoized();
    (Date.now() - start < functionDelayTime).should.be.true;

    // Wait some time, ttl should have expired
    await Promise.delay(ttl + 10);
    start = Date.now();
    await memoized();
    (Date.now() - start >= functionDelayTime).should.be.true;
  });

  it('should allow ttl to be a function', async () => {
    const functionDelayTime = 10;
    const ttl = 100;
    const fn = () => Promise.delay(functionDelayTime);
    const memoized = memoize(fn, {name: 'fn1', ttl: () => ttl});

    let start = Date.now();
    await memoized();
    (Date.now() - start >= functionDelayTime).should.be.true;

    // Call immediately again. Should be a cache hit
    start = Date.now();
    await memoized();
    (Date.now() - start <= functionDelayTime).should.be.true;

    // Wait some time, ttl should have expired;
    await Promise.delay(ttl + 10);

    start = Date.now();
    await memoized();
    (Date.now() - start >= functionDelayTime).should.be.true;
  });

  // FIXME this test is flaky when failing
  it('should gracefully fail lookup timeout', async () => {
    let counter = 0;
    const fn = () => counter++;
    const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
      lookup_timeout: -1,
    });

    const memoized = do_memoize(fn, {name: 'timeout_fn', ttl: 1000});

    // Hack to ensure this takes a while
    const gzipOrig = memoizePkg.gzip;
    memoizePkg.gzip = async function(value) {
      await Promise.delay(10);
      return value;
    };

    const res = await memoized();
    res.should.equal(0);
    const res2 = await memoized();
    res2.should.equal(1); // bc we didn't end up going to redis

    // Restore
    memoizePkg.gzip = gzipOrig;
  });

  it('should work if complex types are accepted and returned', async () => {
    const functionDelayTime = 10;
    const fn = async (arg1) => {
      await Promise.delay(functionDelayTime);
      return { arg1, some: ['other', 'data'] };
    };

    const memoized = memoize(fn, {name: 'fn1'});

    let start = Date.now();
    let { arg1, some } = await memoized({ input: 'data' });
    (Date.now() - start >= functionDelayTime).should.be.true;
    arg1.should.eql({ input: 'data' });
    some.should.eql(['other', 'data']);

    start = Date.now();
    ({ arg1, some } = await memoized({ input: 'data' }));
    (Date.now() - start <= functionDelayTime).should.be.true;
    arg1.should.eql({ input: 'data' });
    some.should.eql(['other', 'data']);
  });

  it('should memoize even if result is falsy', async () => {
    await Promise.all([undefined, null, false, ''].map(async (falsyValue, i) => {
      let callCount = 0;
      const fn = async () => {
        callCount++;
        return falsyValue;
      };

      const memoized = memoize(fn, {name: 'falsy_fn' + i});

      (await memoized() === falsyValue).should.be.true;
      (await memoized() === falsyValue).should.be.true;  // Repeated, presumably cache-hit
      callCount.should.equal(1);  // Verify cache hit
    }));
  });

  it(`shouldn't memoize errors by default`, async () => {
    let callCount = 0;
    const fn = async () => {
      callCount++;
      throw new Error('Test error');
    };

    const memoized = memoize(fn, {name: 'fn1'});

    try {
      await memoized();
    } catch(e) {
      callCount.should.equal(1);
    }

    try {
      await memoized();
    } catch(e) {
      callCount.should.equal(2);
    }
  });

	it(`should restore Date objects, not strings`, async () => {
		const date1 = new Date("2000-01-01T00:00:00.000Z");
		const date2 = new Date("2000-01-02T00:00:00.000Z");

		const fn = async (arg1) => {
			await Promise.delay(500);
			return [arg1, date2];
		};
		const memoized = memoize(fn, {name: 'fn1'});

		let start = Date.now();
		let [val1, val2] = await memoized(date1);
		(Date.now() - start >= 500).should.be.true;
		val1.should.be.an.instanceOf(Date);
		val2.should.be.an.instanceOf(Date);
		val1.should.eql(date1);
		val2.should.eql(date2);

		start = Date.now();
		[val1, val2] = await memoized(date1);
		(Date.now() - start <= 100).should.be.true;
		val1.should.be.an.instanceOf(Date);
		val2.should.be.an.instanceOf(Date);
		val1.should.eql(date1);
		val2.should.eql(date2);
	});

	it('should memoize errors when indicated', async function() {
		let hits = 0;
		async function errorFunction() {
			hits++;
			if (hits === 1) throw new Error('Hit Error #' + hits);
		}

		const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
			memoize_errors_when: () => true,
		});

		const memoized = do_memoize(errorFunction, {name: 'fn1', ttl: 1000});
		await memoized().should.be.rejectedWith(Error, { message: 'Hit Error #1' });
		await memoized().should.be.rejectedWith(Error, { message: 'Hit Error #1' });
	});

	it('should not memoize errors that are filtered out', async function() {
		let hits = 0;
		function errorFunction() {
			hits++;
			if (hits === 1) throw new Error('Hit Error!');
			else if (hits === 2) throw new Error('Special Error!');
			return hits;
		}

		const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
			memoize_errors_when: function(err) {
				return err.message !== 'Hit Error!';
			}
		});

		// First error won't be memoized, so expect hits to increment on subsequent calls.
		// Second error will, so hits will freeze there.
		const memoized = do_memoize(errorFunction, {name: 'fn2', ttl: 1000});
		await memoized().should.be.rejectedWith(Error, { message: 'Hit Error!' });
		hits.should.eql(1);

		await memoized().should.be.rejectedWith(Error, { message: 'Special Error!' });
		hits.should.eql(2);

		// Should stick here as it was memoized
		await memoized().should.be.rejectedWith(Error, { message: 'Special Error!' });
		hits.should.eql(2);
	});

	it('should not memoize two identical-looking functions to the same key', async function() {
		const funcA = (function() {
			const a = 10;
			return async () => a;
		})();
    funcA._name = 'a';
		const funcB = (function() {
			const a = 20;
			return async () => a;
		})();
    funcB._name = 'b';

		const memoizedA = memoize(funcA, {});
		const memoizedB = memoize(funcB, {});
		const result1 = await memoizedA();
		result1.should.eql(10);

		const result2 = await memoizedB();
		result2.should.eql(20);
	});

  it('Should allow overriding error_serialization_keys', async function() {
    let counter = 0;
    async function funcA() {
      const err = new Error('wat');
      err.foo = counter++;
      throw err;
    }

    const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
      memoize_errors_when: function(err) {
        return true;
      },
      error_serialization_keys: ['message', 'foo']
    });

    const memoizedA = do_memoize(funcA, {name: 'err_reviver', ttl: 10000});

    async function verify() {
      try {
        await memoizedA();
      } catch (e) {
        e.message.should.eql('wat');
        e.foo.should.eql(0);
      }
    }
    await verify();
    await verify();
  });

  it('Should allow overriding serialize_value', async function() {
    let counter = 0;
    async function funcA() {
      return counter++;
    }

    const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
      serialize_value(value, options) {
        return `"foo${value}"`;
      },
    });

    const memoizedA = do_memoize(funcA, {name: 'err_reviver', ttl: 10000});

    const value = await memoizedA();
    value.should.eql(0);
    const value2 = await memoizedA();
    value2.should.eql('foo0');
  });

  it('Should allow overriding deserialize_value', async function() {
    let counter = 0;
    async function funcA() {
      return counter++;
    }

    const do_memoize = memoizePkg(client, {
      ...makeDefaultOptions(),
      deserialize_value(value, options) {
        return `bar${value}`;
      },
    });

    const memoizedA = do_memoize(funcA, {name: 'err_reviver', ttl: 10000});

    const value = await memoizedA();
    value.should.eql(0);
    const value2 = await memoizedA();
    value2.should.eql('bar0');
  });
});


index.js
'use strict';
const crypto = require('crypto');
const zlib = require('zlib');
const util = require('util');
const makeLockFn = require('./lock');
const Promise = require('bluebird');
const {clientTyp, exec, isReady} = require('./redisCompat');

const GZIP_MAGIC = new Buffer('$gzip__');
const MAGIC = {
  gzip: GZIP_MAGIC,
  gzip_length: GZIP_MAGIC.length,
  undefined: '_$$_undefined',
  null: '_$$_null',
  not_found: '_$$_empty',
  error: '_$$_error',
};

// Used as filter function in JSON.parse so it properly restores dates
const reISO = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2}(?:\.\d*))(?:Z|(\+|-)([\d|:]*))?$/;

const defaultOptions = {
  // Properties prefixed with `default_` can be overridden when creating each memoized function.
  // How long to persist memoized results to Redis. This can be overridden per-fn.
  default_ttl: 120000,
  // How long to wait for the lock (including retries)
  default_lock_timeout: 5000,

  // How long to wait on Redis before just moving on.
  // If the TTL fed to `memoize` is shorter than this, it will be used instead.
  lookup_timeout: 1000, // ms
  // Set to a function that determines whether or not to memoize an error. By default, we never do.
  memoize_errors_when: (err) => false,
  // Namespace to use under `memos:` in redis. Useful to e.g. invalidate all cache after
  // an application update; simply set this to the current git revision or use the boot timestamp.
  memoize_key_namespace: null,
  // How often to spin on the lock
  lock_retry_delay: 50,
  // Error logger, arity is (err, client, key)
  on_error: null, // *must* be implemented
  // Used for reviving JSON values
  deserialize_value: defaultDeserializeValue,
  serialize_value: defaultSerializeValue,
  error_serialization_keys: ['name', 'stack'],
};

function createMemoizeFunction(client, options = {}) {
  const typ = clientTyp(client);
  options = Object.assign({}, defaultOptions, options);

  // Allow custom namespaces, e.g. by git revision.
  options.keyNamespace = `memos${options.memoize_key_namespace ? ':' + options.memoize_key_namespace : ''}`;
  const lock = makeLockFn(client, options.lock_retry_delay);

  // Validation
  try {
    if (!typ) {
      throw new Error('Pass a Redis client as the first argument.');
    } else if (typ === 'node_redis') {
      if (!client.getAsync) {
        throw new Error('Node_Redis clients must be promisified. Please use Bluebird to do this.');
      }
      if (!client.options.return_buffers) {
        throw new Error('A Node_Redis client passed to the memoizer must have the option `return_buffers` set to true.');
      }
    }

    if (options.error_logging) {
      throw new Error('The "error_logging" var has been replaced with a more standard ' +
        '"on_error(err, client, key) callback." in 4.0');
    }
    if (typeof options.on_error !== 'function') {
      throw new Error('An error fn of the arity (err, client, key) must be passed as an option.');
    }
  } catch (e) {
    e.message = `Redis-Memoizer: ${e.message}`;
    throw e;
  }

  return memoizeFn.bind(null, client, options, lock);
}

function getFunctionKey(fn, name = fn._name) {
  if (!name) throw new Error("Unable to determine memoization name for function: " + fn);
  return name;
}

function hash(args) {
  return crypto.createHash('sha1').update(JSON.stringify(args)).digest('hex');
}

function memoizeFn(client, options, lock, fn,
                   {ttl = options.default_ttl, lock_timeout = options.default_lock_timeout, name} = {}) {
  let functionKey = module.exports.getFunctionKey(fn, name);
  const ttlfn = typeof ttl === 'function' ? ttl : () => ttl;

  return async function memoizedFunction(...args) {
    // Hash the args so we can look for this key in redis.
    const argsHash = module.exports.hash(args);

    // Set a timeout on the retrieval from redis.
    const timeoutMs = Math.min(ttlfn(), options.lookup_timeout);
    const key = `${options.keyNamespace}:${functionKey}:${argsHash}`;

    // Attempt to get the result from redis.
    const memoValue = await doLookup(client, key, timeoutMs, options);
    // We return an internal marker if this thing was actually not found, versus just null
    if (memoValue !== MAGIC.not_found) return memoValue;

    // Ok, we're going to have to actually execute the function.
    // Lock ensures only one fn executes at a time and prevents a stampede.
    const unlock = await lock(key, lock_timeout);
    let didOriginalFn = false;
    try {
      // After we've acquired the lock, check if the key was populated in the meantime.
      const memoValueRetry = await doLookup(client, key, timeoutMs, options);
      if (memoValueRetry !== MAGIC.not_found) return memoValueRetry;

      // Run the fn, save the result
      didOriginalFn = true;
      const result = await fn.apply(this, args);
      // Write the key, but don't await on it
      writeKeyToRedis(client, key, result, ttlfn(result), options)
      .catch((err) => {
        err.message = `Redis-Memoizer: Error writing key "${key}": ${err.message}`;
        options.on_error(err, client, key);
      });

      return result;
    } catch (e) {
      // original function errored, should we memoize that?
      if (didOriginalFn && options.memoize_errors_when(e)) {
        await writeKeyToRedis(client, key, e, ttlfn(e), options);
      }
      throw e;
    } finally {
      unlock();
    }
  };
}

async function doLookup(client, key, timeout, options) {
  let memoValue;
  try {
    memoValue = await Promise.resolve(getKeyFromRedis(client, key, options)).timeout(timeout);
  } catch (err) {
    err.message = `Redis-Memoizer: Error getting key "${key}" with timeout ${timeout}: ${err.message}`;
    if (err.name !== 'TimeoutError') {
      options.on_error(err, client, key);
    }
    // Continue on
    return MAGIC.not_found;
  }
  if (memoValue instanceof Error) throw memoValue; // we memoized an error.
  return memoValue;
}

async function getKeyFromRedis(client, key, options) {
  // Bail if not connected; don't wait for reconnect, that's probably slower than just computing.
  if (!isReady(client)) throw new Error('Not connected.');

  let value = await compressedGet(client, key);

  // Coerce back
  if (value instanceof Buffer) value = value.toString(); // redis/ioredis compat
  if (value == null) return MAGIC.not_found;
  return options.deserialize_value(value, options);
}

function defaultDeserializeValue(value, options) {
  if (value === MAGIC.undefined) return undefined;
  else if (value === MAGIC.null) return null;
  else if (value === '') return value;
  else return JSON.parse(value, reviver.bind(null, options));
}

function reviver(options,key, value) {
  // Revive dates
  if (typeof value === 'string' && reISO.exec(value)) {
    return new Date(value);
  }
  // Revive errors
  else if (value && value.hasOwnProperty(MAGIC.error)) {
    const err = new Error(value.message);
    for (let i = 0; i < options.error_serialization_keys.length; i++) {
      const key = options.error_serialization_keys[i];
      err[key] = value[key];
    }
    return err;
  }
  return value;
}

function defaultSerializeValue(value, options) {
  // Convert a couple of placeholder values so we can distinguish between a value that is null/undefined,
  // and the key not found.
  let serializedValue;
  if (value === undefined) {
    serializedValue = MAGIC.undefined;
  } else if (value === null) {
    serializedValue = MAGIC.null;
  } else if (value instanceof Error) {
    // Mark errors so we can revive them
    value[MAGIC.error] = true;
    // Seems to do pretty well on errors
    serializedValue = JSON.stringify(value, options.error_serialization_keys.concat(['message', MAGIC.error]));
  } else {
    serializedValue = JSON.stringify(value);
  }
  return serializedValue;
}

async function writeKeyToRedis(client, key, value, ttl, options) {
  if (!isReady(client)) throw new Error('Not connected.');

  // Don't bother writing if ttl is 0.
  if (ttl === 0) return;

  const serializedValue = options.serialize_value(value, options);
  return compressedPSetX(client, key, ttl, serializedValue);
}

//
// GZIP
//

const gzipAsync = util.promisify(zlib.gzip);
const gunzipAsync = util.promisify(zlib.gunzip);

async function compressedGet(client, key, cb) {
  let zippedVal;
  // Have to use 'getBuffer' for ioredis
  if (clientTyp(client) === 'ioredis') zippedVal = await client.getBuffer(key);
  else zippedVal = await client.getAsync(key);
  return module.exports.gunzip(zippedVal);
}

async function compressedPSetX(client, key, ttl, value) {
  const zippedVal = await module.exports.gzip(value);
  return exec(client, 'set', key, zippedVal, 'PX', ttl);
}

async function gzip(value) {
  // null or too small to effectively gzip
  if (value == null || value.length < 500) return value;

  const zippedVal = await gzipAsync(value);
  return Buffer.concat([MAGIC.gzip, zippedVal], zippedVal.length + MAGIC.gzip_length);
}

async function gunzip(value) {
  // Check for GZIP MAGIC, if there unzip it.
  if (value instanceof Buffer && value.slice(0, MAGIC.gzip_length).equals(MAGIC.gzip)) {
    return gunzipAsync(value.slice(MAGIC.gzip_length));
  } else {
    return value;
  }
}

module.exports = createMemoizeFunction;
// Exported so they can be overridden
module.exports.hash = hash;
module.exports.getFunctionKey = getFunctionKey;
module.exports.gzip = gzip;
module.exports.gunzip = gunzip;
module.exports.MAGIC = MAGIC;
module.exports.reISO = reISO;
